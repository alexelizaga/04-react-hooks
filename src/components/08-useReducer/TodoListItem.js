import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ index, todo, handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={`${todo.done && "complete"}`}
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        onClick={() => handleDelete(todo.id)}
      >
        Borrar
      </button>
    </li>
  )
};

TodoListItem.propTypes = {
  index: PropTypes.number.isRequired,
  todo: PropTypes.object.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}
