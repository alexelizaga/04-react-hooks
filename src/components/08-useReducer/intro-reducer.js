
const initialState = [{
  id: 1,
  todo: 'Comprar pan',
  done: false,
}];

const todoReducer = ( state = initialState, action ) => {
  switch (action?.type) {
    case 'add':
      return [ ...state, action.payload ]
  
    default:
      return state;
  }
}
let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'Comprar leche',
  done: false,
}

const action = { type: 'add', payload: newTodo }

todos = todoReducer( todos, action );






console.log(todos);