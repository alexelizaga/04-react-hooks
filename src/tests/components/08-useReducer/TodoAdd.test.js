import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Test <TodoAdd />', () => {
  const props = {
    handleAddTodo: jest.fn(),
  }
  test('should render correctly', () => {
    const tree = renderer.create(<TodoAdd {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should not call handleAddTodo', () => {
    render(<TodoAdd {...props} />);

    const formSubmit = screen.getByTestId('TodoAddForm');

    fireEvent.submit(formSubmit);

    expect(props.handleAddTodo).toBeCalledTimes(0);

  });

  test('should call handleAddTodo', () => {
    render(<TodoAdd {...props} />);

    const input = screen.getByTestId('TodoAddInput');
    fireEvent.change(input, {
      target: {
        value: 'Aprende React',
        name: 'description',
      }
    });
    expect(input.getAttribute('value')).toBe('Aprende React');

    const formSubmit = screen.getByTestId('TodoAddForm');
    fireEvent.submit(formSubmit, {});
    expect(props.handleAddTodo).toBeCalledTimes(1);
    expect(props.handleAddTodo).toBeCalledWith(expect.any(Object));
    expect(props.handleAddTodo).toBeCalledWith({
      desc: "Aprende React",
      done: false,
      id: expect.any(Number),
    });
    expect(input.getAttribute('value')).toBe('');

  });


});