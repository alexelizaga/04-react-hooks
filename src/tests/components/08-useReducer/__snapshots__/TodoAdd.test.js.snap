// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoAdd /> should render correctly 1`] = `
Array [
  <h4>
    Agregar TODO
  </h4>,
  <hr />,
  <form
    data-testid="TodoAddForm"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      data-testid="TodoAddInput"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>,
]
`;
